edgelet/iotedged/src/lib.rs

```diff
@@ -484,63 +624,137 @@ where
     Ok(())
 }
 
+fn check_x509_reprovisioning<C>(
+    subdir_path: &PathBuf,
+    settings: &Settings<DockerConfig>,
+    crypto: &C,
+    hybrid_id_filename: &str,
+    iv_filename: &str,
+) -> (bool, Option<Vec<u8>>)
+where
+    C: CreateCertificate + Decrypt,
+{
+    fn check_x509_reprovisioning_inner<C>(
+        subdir_path: &PathBuf,
+        crypto: &C,
+        hybrid_id_filename: &str,
+        iv_filename: &str,
+    ) -> Result<(bool, Option<Vec<u8>>), HybridAuthError>
+    where
+        C: CreateCertificate + Decrypt,
+    {
+        // check if the identity key & iv files exist and are valid
+        let key_path = subdir_path.join(hybrid_id_filename);
+        let enc_identity_key = fs::read(key_path.as_path())?;
+        let iv_path = subdir_path.join(iv_filename);
+        let iv = fs::read(iv_path.as_path())?;
+        if iv.len() == IOTEDGED_CRYPTO_IV_LEN_BYTES {
+            let identity_key = crypto.decrypt(
+                IOTEDGED_CRYPTO_ID.as_bytes(),
+                enc_identity_key.as_ref(),
+                &iv,
+            )?;
+            if identity_key.as_ref().len() == IDENTITY_MASTER_KEY_LEN_BYTES {
+                Ok((false, Some(identity_key.as_ref().to_vec())))
+            } else {
+                Ok((true, None))
+            }
+        } else {
+            Ok((true, None))
+        }
+    }
+
+    if get_provisioning_auth_method(settings) == ProvisioningAuthMethod::X509 {
+        check_x509_reprovisioning_inner(subdir_path, crypto, hybrid_id_filename, iv_filename)
+            .unwrap_or((true, None))
+    } else {
+        (false, None)
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
 fn check_settings_state<M, C>(
-    subdir_path: PathBuf,
-    filename: &str,
+    subdir_path: &PathBuf,
```

Comment:
Should be `subdir_path: &Path`. `&PathBuf` is unnecessary double indirection compared to `&Path` (just like `&String` vs `&str`).

Reply:
Cleaned up here and elsewhere.